uast.cpp

```diff
@@ -0,0 +1,119 @@
+#include "uast.h"
+#include "roles.h"
+
+#define GET(__NODE__, __PROP__, ...) (impl.__PROP__(__NODE__, ##__VA_ARGS__))
+
+#define check_ptr(__V__)                                     \
+  if (__V__ == NULL)                                         \
+  {                                                          \
+    throw std::runtime_error("missing node implementation"); \
+  }
+
+UAST::UAST(NodeImpl impl) : impl(impl)
+{
+  checkImpl();
+  xmlInitParser();
+}
+
+UAST::~UAST()
+{
+  xmlCleanupParser();
+}
+
+void UAST::checkImpl()
+{
+  check_ptr(impl.internal_type);
+  check_ptr(impl.token);
+  check_ptr(impl.children_size);
+  check_ptr(impl.child);
+  check_ptr(impl.roles_size);
+  check_ptr(impl.role);
+}
+
+std::vector<void *> UAST::find(void *node, std::string query)
+{
+  auto doc = create_document(node);
+  auto ctx = create_xpath_context(doc);
+  auto xpathObj = evaluate_xpath(ctx, query);
+
+  xmlNodeSetPtr result = xpathObj->nodesetval;
+  int size = (result) ? result->nodeNr : 0;
+  auto nodes = result->nodeTab;
+
+  std::vector<void *> uast;
+  uast.reserve(size);
+  for (int i = 0; i < size; i++)
+  {
+    uast.push_back(nodes[i]->_private);
+  }
+
+  xmlFreeDoc(doc);
+  xmlXPathFreeContext(ctx);
+  xmlXPathFreeObject(xpathObj);
+
+  return move(uast);
+}
+
+xmlDocPtr UAST::create_document(void *node)
+{
+  auto doc = xmlNewDoc(BAD_CAST("1.0"));
+  auto xmlNode = recursive_conv(node, NULL);
+  xmlDocSetRootElement(doc, xmlNode);
+  return doc;
+}
+
+xmlXPathContextPtr UAST::create_xpath_context(xmlDocPtr doc)
+{
+  xmlXPathContextPtr ctx = xmlXPathNewContext(doc);
+  if (ctx == NULL)
+  {
+    throw std::runtime_error("unable to create new XPath context");
```

Comment:
These exceptions should be handled in the library context. Only error values should be returned to the library user, otherwise you're forcing the user to be able to handle C++ exceptions.